### PR TITLE
栈的python代码

### PR DESCRIPTION
**Pull request by pengchzn**

### CODE REVIEW

File: codes/python/chapter_stack_and_queue/array_stack.py [hunk 1]
```diff
@@ -4,7 +4,78 @@
 Author: Krahets (krahets@163.com)
```
Comment (krahets) [inline codes/python/chapter_stack_and_queue/array_stack.py:4]: Please update the author and time information in the file header with yours, thanks!

File: codes/python/chapter_stack_and_queue/array_stack.py [hunk 2]
```diff
@@ -4,7 +4,78 @@
...
+""" 基于数组实现的栈 """
+
+
+class ArrayStack:
+    def __init__(self):
+        self._stack = []
+        self._size = 0
+
+    """ 获取栈的长度 """
+
```
Comment (krahets) [inline codes/python/chapter_stack_and_queue/array_stack.py:22]: No empty line between the function name and the function

File: codes/python/chapter_stack_and_queue/array_stack.py [hunk 3]
```diff
@@ -4,7 +4,78 @@
...
 
-import sys, os.path as osp
+import os.path as osp
+import sys
+
 sys.path.append(osp.dirname(osp.dirname(osp.abspath(__file__))))
 from include import *
 
+""" 基于数组实现的栈 """
+
```
Comment (krahets) [inline codes/python/chapter_stack_and_queue/array_stack.py:14]: No empty line here

**Review by krahets — COMMENTED** [review]
Thanks, @pengchzn ! Some comments need to be addressed.

**Comment by pengchzn** [comment]
Got it, I've already corrected them, thanks!

File: codes/python/chapter_stack_and_queue/array_stack.py [hunk 4]
```diff
@@ -1,10 +1,73 @@
...
+    pop = stack.pop()
+    print("出栈元素 pop = ", pop)
+    print("出栈后 stack = ", stack._stack)
+
+    """ 获取栈的长度 """
+    size = stack.size()
+    print("栈的长度 size = ", size)
+
+    """ 判断是否为空 """
+    isEmpty = (stack == [])
```
Comment (krahets) [inline codes/python/chapter_stack_and_queue/array_stack.py:73]: This should be `isEmpty = stack.isEmpty()`.

And `stack` is a ArrayStack object, which will never satisfy `stack == []`

File: codes/python/chapter_stack_and_queue/array_stack.py [hunk 5]
```diff
@@ -1,10 +1,73 @@
...
+    def __init__(self):
+        self._stack = []
+        self._size = 0
+
+    """ 获取栈的长度 """
+    def size(self):
+        return self._size
+
+    """ 判断栈是否为空 """
+    def isEmpty(self):
```
Comment (krahets) [inline codes/python/chapter_stack_and_queue/array_stack.py:24]: is_emtpy()

File: codes/python/chapter_stack_and_queue/linkedlist_stack.py [hunk 6]
```diff
@@ -1,10 +1,84 @@
...
+
+    """ 访问栈顶元素 """
+    def peek(self):
+        return self.head.val
+
+    def to_array(self):
+        stack = []
+        temp = self.head
+        while temp:
+            stack.insert(0, temp.val)
```
Comment (krahets) [inline codes/python/chapter_stack_and_queue/linkedlist_stack.py:53]: It is recommended `append(temp.val)` here, and followed by  `temp.reverse()` to reduce time complexity.

File: codes/python/chapter_stack_and_queue/linkedlist_stack.py [hunk 7]
```diff
@@ -1,10 +1,84 @@
...
+    pop = stack.pop()
+    print("出栈元素 pop = ", pop)
+    print("出栈后 stack = ", stack.to_array())
+
+    """ 获取栈的长度 """
+    size = stack.size()
+    print("栈的长度 size = ", size)
+
+    """ 判断是否为空 """
+    isEmpty = (stack == [])
```
Comment (krahets) [inline codes/python/chapter_stack_and_queue/linkedlist_stack.py:84]: stack will never satisfy `stack == []` You should use `stack.is_empty()`

**Review by krahets — CHANGES_REQUESTED** [review]
A few more comments, thanks!

File: codes/python/chapter_stack_and_queue/linkedlist_stack.py (hunk 6 above)
Comment (pengchzn) [reply codes/python/chapter_stack_and_queue/linkedlist_stack.py:53]: The ListNode type seems doed not have the function of reverse(), so I slice the index of it instead, that should be faster than using reverse

Comment (pengchzn) [reply codes/python/chapter_stack_and_queue/linkedlist_stack.py:53]: Got it, I slice the index of it instead, for it should be faster than using reverse(),thanks for comments

**Review by krahets — APPROVED** [review]

**Comment by krahets** [comment]
Thanks for the code! @pengchzn 

By the way, `reverse()` is better than slicing to reverse a list. More information at

https://www.geeksforgeeks.org/python-reversed-vs-1-which-one-is-faster/

**Comment by pengchzn** [comment]
Ah, I will check that article for more information, maybe I keep the wrong idea for quite a long time. Thanks for that

**Comment by krahets** [comment]
> Ah, I will check that article for more information, maybe I keep the wrong idea for quite a long time. Thanks for that

https://github.com/krahets/hello-algo/commit/53cc651af2ac03f69f7b68d21fbecbbda747cf85

I update the implementation of the array stack, especially the linked list stack. Please check it.

**Pull request merged**